libs/langchain/langchain/vectorstores/cratedb/base.py

```diff
@@ -0,0 +1,396 @@
+            texts=texts, embeddings=embeddings, metadatas=metadatas, ids=ids, **kwargs
+        )
+
+    def create_tables_if_not_exists(self) -> None:
+        """
+        Need to overwrite because `Base` is different from upstream.
+        """
+        Base.metadata.create_all(self._engine)
+
+    def drop_tables(self) -> None:
+        """
+        Need to overwrite because `Base` is different from upstream.
+        """
+        Base.metadata.drop_all(self._engine)
+
+    def delete(
+        self,
+        ids: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Delete vectors by ids or uuids.
+
+        Remark: Specialized for CrateDB to synchronize data.
+
+        Args:
+            ids: List of ids to delete.
+
+        Remark: Patch for CrateDB needs to overwrite this, in order to
+                add a "REFRESH TABLE" statement afterwards. The other
+                patch, listening to `after_delete` events seems not be
+                able to catch it.
+        """
+        super().delete(ids=ids, **kwargs)
+
+        # CrateDB: Synchronize data because `on_flush` does not catch it.
+        with self.Session() as session:
+            refresh_table(session, self.EmbeddingStore)
+
+    @property
+    def distance_strategy(self) -> Any:
+        if self._distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return self.EmbeddingStore.embedding.euclidean_distance
+        elif self._distance_strategy == DistanceStrategy.COSINE:
+            raise NotImplementedError("Cosine similarity not implemented yet")
+        elif self._distance_strategy == DistanceStrategy.MAX_INNER_PRODUCT:
+            raise NotImplementedError("Dot-product similarity not implemented yet")
+        else:
+            raise ValueError(
+                f"Got unexpected value for distance: {self._distance_strategy}. "
+                f"Should be one of {', '.join([ds.value for ds in DistanceStrategy])}."
+            )
+
+    def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, float]]:
+        """Return docs and scores from results."""
+        docs = [
+            (
+                Document(
+                    page_content=result.EmbeddingStore.document,
+                    metadata=result.EmbeddingStore.cmetadata,
+                ),
+                result._score if self.embedding_function is not None else None,
+            )
+            for result in results
+        ]
+        return docs
+
+    def _query_collection(
+        self,
+        embedding: List[float],
+        k: int = 4,
+        filter: Optional[Dict[str, str]] = None,
+    ) -> List[Any]:
+        """Query the collection."""
+        with self.Session() as session:
+            collection = self.get_collection(session)
+            if not collection:
+                raise ValueError("Collection not found")
+
+            filter_by = self.EmbeddingStore.collection_id == collection.uuid
```

Comment:
About the question how working with different collections works with LangChain. /cc @ckurze